COMOPT-2359 Document Private Catalog Views and Restricted Access Keys for GraphQL (#354)
AdobeDocs/commerce-services: src/pages/optimizer/merchandising-services/using-the-api.md

Author: ```suggestion
```

# Get started with the Merchandising API

Use the Merchandising API to retrieve product data from your Commerce catalogs and display it in Commerce frontend experiences. Data includes products, categories, product and category attribute metadata, price books, and prices.

## Prerequisites

Before using the Merchandising API, ensure you have:

> - **Adobe Commerce Optimizer Access**: Active subscription and the instance ID associated with your Adobe Commerce Optimizer instance

Author: ```suggestion
- **Adobe Commerce Optimizer access**: Active subscription and the instance ID associated with your Adobe Commerce Optimizer instance
```

> - **Catalog Data**: Products and pricing data ingested via the [Data Ingestion API](../data-ingestion/index.md)

Author: ```suggestion
- **Catalog data**: Products and pricing data ingested via the [Data Ingestion API](../data-ingestion/index.md)
```

> - **Catalog views**: Configured views and policies in Adobe Commerce Optimizer

Reviewer: ```suggestion
- **Catalog Views**: Configured views and policies in Adobe Commerce Optimizer
```

> - **Catalog views**: Configured views and policies in Adobe Commerce Optimizer

Author: ```suggestion
- **Catalog views**: Configured views and policies in [Adobe Commerce Optimizer Studio](https://experienceleague.adobe.com/en/docs/commerce/optimizer/setup/catalog-view)
```

### Authentication

Authentication is not required for the Merchandising API by default.

> However, requests for data from private catalog views configured with catalog protection and restricted access keys must include a valid, signed JSON Web Token (JWT) in the `AC-Catalog-View-Access-Token` header.

Reviewer: Do you need to update the list of headers? (Not sure if it's an include file.)

Author: No-- the headers are auto-generated (includes/autogenerated)

> Setting up a private catalog view is the responsibility of your client application:

Author: ```suggestion
You can set up a private catalog view in Adobe Commece Optimizer Studio by enabling catalog protection and assigning a restricted access key to a catalog view.

- **Generate an RSA key pair.** The public key must be PEM-encoded and between 2048 and 8192 bits.

> - **Register the public key** as a restricted access key on the catalog view. See [Restricted access keys](https://experienceleague.adobe.com/en/docs/commerce/optimizer/setup/restricted-access-keys) and [Protect a catalog view](https://experienceleague.adobe.com/en/docs/commerce/optimizer/setup/catalog-view#protect-a-catalog-view).

Author: ```suggestion
- **Register the public key** as a restricted access key on the catalog view. See [Restricted access keys](https://experienceleague.adobe.com/en/docs/commerce/optimizer/setup/restricted-access-keys) and [Protect a catalog view](https://experienceleague.adobe.com/en/docs/commerce/optimizer/setup/private-catalog-view#protect-a-catalog-view).
```

Adobe Commerce Optimizer validates each token's `RS256` signature against the Restricted Access Keys assigned to the catalog view, and returns catalog data only if the signature is valid and neither the token nor the key has expired.

![Sequence diagram showing the Private Catalog View authentication flow: registering an RSA public key as a Restricted Access Key, then signing and validating a JWT on each request](../../images/merchandising/adobe-commerce-optimizer-auth-sequence.png)

A request to a Private Catalog View without a valid token returns a GraphQL error instead of data:

```json
{
  "errors": [
    {
      "message": "Access key validation failed: Missing token",
      "extensions": { "x-commerce-exception": "access-key-invalid" }
    }
  ]
}
```

The `message` field describes why validation failed:

| Reason | Cause |
|---|---|
| `Missing token` | No `AC-Catalog-View-Access-Token` header was sent on a Private Catalog View. |
| `Access token signature invalid` | The token's signature doesn't verify against any Restricted Access Key assigned to the catalog view, for example because it was signed with the wrong key or was tampered with. |

> | `Restricted access is enabled but no valid access keys are available` | Catalog Protection is enabled, but no Restricted Access Keys are assigned to the catalog view, or every assigned key has expired. |

Author: ```suggestion
   | `accessToken` | Required for private catalog views. The signed JWT proving authorization to access the catalog view specified by `catalogViewId`, for example `eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9...`.|
```

> A token that has passed its own expiration (the `exp` claim) is denied the same way, even if the signature would otherwise validate.

Reviewer: `exp` wasn't previously mentioned

> A token that has passed its own expiration (the `exp` claim) is denied the same way, even if the signature would otherwise validate.

Author: ```suggestion
A token that has passed its own expiration date is denied the same way, even if the signature would otherwise validate.
```

### Headers

> When making requests to the Merchandising API, you must include required HTTP headers that provide necessary information, such as the catalog view ID, catalog source locale. Additionally, you can include optional policy and price book headers that tailor the API response to your needs.

Author: ```suggestion
When making requests to the Merchandising API, you must include required HTTP headers that provide necessary information, such as the catalog view ID, catalog source locale. Additionally, you can include optional policy, price book, or catalog access view token headers that tailor the API response to your needs.
```

--header 'AC-View-ID: {{catalogViewId}}'  \
--header 'AC-Price-Book-ID: {{priceBookId}}'  \

> --header 'AC-Policy-{{attributeCode}}: {{attributeValue}}'  \

Author: ```suggestion
  --header 'AC-Catalog-View-Access-Token: {{accessToken}}' 
```

     -d '{"query": "query ProductSearch($search: String!) { productSearch( phrase: $search, page_size: 10) { items { productView { sku name description shortDescription images { url } ... on SimpleProductView { attributes { label name value } price { regular { amount { value currency } } roles } } } } } }", "variables": { "search": "your-string"}}'
   ```

   If the catalog view specified by `AC-View-ID` is a Private Catalog View, add the `AC-Catalog-View-Access-Token` header with a valid signed JWT to the request. See [Authentication](#authentication) for how the token is generated and validated.

   ```shell
   curl --request POST \
   --url https://na1-sandbox.api.commerce.adobe.com/{{tenantId}}/graphql \
   --header 'AC-View-ID: {{catalogViewId}}'  \
   --header 'AC-Price-Book-ID: {{priceBookId}}'  \
   --header 'AC-Catalog-View-Access-Token: {{accessToken}}'  \
   --data '{{apiPayload}}'
   ```

   | Placeholder name | Description                                                                                                     |
   |------------------|-----------------------------------------------------------------------------------------------------------------|

>    | `accessToken` | Required for Private Catalog Views. The signed JWT proving authorization to access the catalog view specified by `catalogViewId`, for example `eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9...`.|

Reviewer: this is a one-row table. Either add the other placeholders or remove the table markup
